kernel/thread.c

```diff
@@ -605,12 +606,49 @@ myst_thread_t* myst_find_thread(int tid)
                 break;
             }
         }
-        myst_spin_unlock(thread->thread_lock);
     }
 
+    myst_spin_unlock(thread->thread_lock);
     return target;
 }
 
+/* Find a thread of a specific process */
+myst_thread_t* myst_find_process_specific_thread(pid_t pid, pid_t tid)
+{
+    myst_thread_t* our_process_thread =
+        myst_find_process_thread(myst_thread_self());
+    myst_thread_t* find_thread = our_process_thread;
+
+    myst_spin_lock(&myst_process_list_lock);
+    while (find_thread && find_thread->pid != pid)
+    {
+        find_thread = find_thread->main.prev_process_thread;
+    }
+    if (find_thread == NULL)
+    {
+        find_thread = our_process_thread->main.next_process_thread;
+        while (find_thread && find_thread->pid != pid)
+        {
+            find_thread = find_thread->main.next_process_thread;
+        }
+    }
+    myst_spin_unlock(&myst_process_list_lock);
+
+    if (find_thread == NULL)
+        goto done;
+
+    myst_spin_lock(find_thread->thread_lock);
+    while (find_thread && find_thread->tid != tid)
+    {
+        find_thread = find_thread->group_next;
```

Comment:
Are we sure the process thread is always the first in the thread group list?

Reply:
yes

Reply:
there are a number of assumptions made elsewhere of the fact